### PR TITLE
VBO support

### PR DESCRIPTION
**Pull request by mitchej123**
Adds some basic VBO support, as well as a CapturingTessellator useful for capturing multiple draw() calls from something such as RenderBlockAsItem and converting the quads into a VBO.

### CODE REVIEW

File: src/main/java/com/gtnewhorizons/angelica/rendering/AngelicaBlockSafetyRegistry.java [hunk 1]
```diff
@@ -11,7 +11,7 @@ public class AngelicaBlockSafetyRegistry {
     private static final StampedLock LOCK = new StampedLock();
 
     public static boolean canBlockRenderOffThread(Block block) {
-        long stamp = LOCK.readLock();
+        final long stamp = LOCK.readLock();
```
Comment (embeddedt) [inline src/main/java/com/gtnewhorizons/angelica/rendering/AngelicaBlockSafetyRegistry.java:14]: FYI - I don't think it makes a difference to runtime performance/behavior to have `final` here or not.

Comment (mitchej123) [reply src/main/java/com/gtnewhorizons/angelica/rendering/AngelicaBlockSafetyRegistry.java:14]: It probably doesn't but I have my ide set to complain about non final and do it as I touch things

**Review by makamys — CHANGES_REQUESTED** [review]
Once this is merged we should also yeet `BlockRenderer#tessellatorToQuadList` and use a `CapturingTessellator`. This would allow block renderers that restart the tessellator ([Matter Overdrive comes to mind](https://github.com/makamys/Neodymium/issues/5#issuecomment-1176217533)) to work properly.

File: src/main/java/com/gtnewhorizons/angelica/client/renderer/CapturingTessellator.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+ *
+ * Used to capture the quads generated by the Tessellator across multiple draw calls and make the quad list available for usage.
+ *   Currently being used to generate a VBO for a model and only stores Position, Texture, and Normals.
+ *
+ * NOTE: This should only be used on the main thread, and will _not_ (currently) capture, integrate, or stop any GL calls made
+ * around the tessellator draw calls.
+ */
+@SuppressWarnings("unused")
+public class CapturingTessellator extends Tessellator {
+    protected final VertexFormat format;
```
[the remark this answers is not in the document]
Comment (mitchej123) [reply src/main/java/com/gtnewhorizons/angelica/client/renderer/CapturingTessellator.java:25]: Ah you're right, I ended up reworking it after the first go at it.

File: src/main/java/com/gtnewhorizons/angelica/compat/mojang/VertexBuffer.java [hunk 3]
```diff
@@ -23,6 +26,47 @@ public void unbind() {
...
+        } catch(Exception e) {
+            throw new RuntimeException("Failed to upload vertex buffer", e);
+        } finally {
+            ((ITessellatorInstance) tessellator).discard();
+        }
+    }
+
+    public ByteBuffer tessellatorToBuffer(Tessellator tessellator, VertexFormat format) {
+
+        final int verticesPerPrimitive = tessellator.drawMode == GL11.GL_QUADS ? 4 : 3;
```
[the remark this answers is not in the document]
Comment (mitchej123) [reply src/main/java/com/gtnewhorizons/angelica/compat/mojang/VertexBuffer.java:41]: The plan was to make it work with triangles eventually, probably best to just throw an exception until then

File: src/main/java/com/gtnewhorizons/angelica/glsm/TessellatorManager.java [hunk 4]
```diff
@@ -20,6 +28,18 @@ public static boolean isMainInstance(Object instance) {
         return instance == Tessellator.instance || isOnMainThread();
     }
 
+    public static void startCapturing(CapturingTessellator tessellator) {
```
[the remark this answers is not in the document]
Comment (mitchej123) [reply src/main/java/com/gtnewhorizons/angelica/glsm/TessellatorManager.java:31]: I was originally pondering how to allow this to work in the background threads for ISBRH and potentially TESRs (coupled with a capturing GLSM).  However a thread local would probably work.

**Comment by mitchej123** [comment]
> Once this is merged we should also yeet BlockRenderer#tessellatorToQuadList and use a CapturingTessellator

I went ahead and did this as part of this PR

File: src/main/java/com/gtnewhorizons/angelica/client/renderer/CapturingTessellator.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        super.reset();
+        this.offset.zero();
+    }
+
+    public List<Quad> getQuads() {
+        return quadBuf.getAsList();
+    }
+
+    public void resetQuadBuf() {
+        this.quadBuf.reset();
```
Comment (makamys) [inline src/main/java/com/gtnewhorizons/angelica/client/renderer/CapturingTessellator.java:78]: `clearQuads` would be a more straightforward name, given that this is a counterpart to `getQuads`.

**Review by makamys — APPROVED** [review]

**Pull request merged**